fcrepo-connector-file/src/main/java/org/fcrepo/connector/file/ExternalJsonSidecarExtraPropertyStore.java

```diff
@@ -0,0 +1,171 @@
+ */
+package org.fcrepo.connector.file;
+
+import org.fcrepo.kernel.exception.RepositoryRuntimeException;
+import org.infinispan.schematic.Schematic;
+import org.infinispan.schematic.document.Document;
+import org.infinispan.schematic.document.EditableDocument;
+import org.infinispan.schematic.document.Json;
+import org.modeshape.jcr.cache.document.DocumentTranslator;
+import org.modeshape.jcr.spi.federation.ExtraPropertiesStore;
+import org.modeshape.jcr.value.Name;
+import org.modeshape.jcr.value.Property;
+
+import java.util.Collections;
+import java.io.File;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.io.IOException;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+
+/**
+ * An implementation of ExtraPropertyStore, based on
+ * org.modeshape.connector.filesystem.JsonSidecarExtraPropertyStore that stores the
+ * properties in a separate configured directory than the filesystem federation itself.
+ *
+ * @author Mike Durbin
+ * @author acoburn
+ */
+public class ExternalJsonSidecarExtraPropertyStore implements ExtraPropertiesStore {
+
+    private FedoraFileSystemConnector connector;
+
+    private File propertyStoreRoot;
+
+    private final DocumentTranslator translator;
+
+    /**
+     * Default constructor.
+     * @param connector the FileSystemConnector for which this class will store properties.
+     * @param propertyStoreRoot the root of a filesystem into which properties will be
+     *                          serialized.
+     */
+    public ExternalJsonSidecarExtraPropertyStore(final FedoraFileSystemConnector connector,
+                                                 final DocumentTranslator translator,
+                                                 final File propertyStoreRoot) {
+        this.connector = connector;
+        this.translator = translator;
+        this.propertyStoreRoot = propertyStoreRoot;
+    }
+
+    protected File sidecarFile(final String id) {
+        final File file;
+        if (connector.isRoot(id)) {
+            file = new File(propertyStoreRoot, "federation-root.modeshape.json");
+        } else {
+            String ext = ".modeshape.json";
+            if (connector.isContentNode(id)) {
+                ext = ".content.modeshape.json";
+            }
+            final File f = new File(connector.fileFor(id).getAbsolutePath() + ext);
+
+            final Path propertyFileInFederation = f.getAbsoluteFile().toPath();
+            final Path relativePath = connector.fileFor("/")
+                            .getAbsoluteFile().toPath().relativize(propertyFileInFederation);
+            file = propertyStoreRoot.toPath().resolve(relativePath).toFile();
+        }
+        if (!file.getParentFile().exists() && !file.getParentFile().mkdirs()) {
+            throw new RepositoryRuntimeException("Unable to create directories " + file.getParentFile() + ".");
+        }
+        return file;
+    }
+
+    @Override
```

Comment:
For the methods that you copied from ModeShape's codebase, could you provide links to the ModeShape GitHub source so we can check for updates in the future.